ui/src/app/workflows/components/submit-workflow-panel.tsx

```diff
@@ -17,119 +17,92 @@ interface Props {
-                    <div key='entrypoint' title='Entrypoint' style={{marginBottom: 25}}>
-                        <label>Entrypoint</label>
-                        <Select
-                            value={this.state.entrypoint}
-                            options={this.state.templates.map(t => ({
-                                value: t.name,
-                                title: t.name
-                            }))}
-                            onChange={selected => {
-                                const selectedTemplate = this.getSelectedTemplate(selected.value);
-                                this.setState({
-                                    entrypoint: selected.value,
-                                    selectedTemplate,
-                                    parameters: (selectedTemplate && selectedTemplate.inputs.parameters) || []
-                                });
-                            }}
-                        />
-                    </div>
-                    <div key='parameters' style={{marginBottom: 25}}>
-                        <label>Parameters</label>
-                        {this.state.workflowParameters.length > 0 && (
-                            <ParametersInput parameters={this.state.workflowParameters} onChange={workflowParameters => this.setState({workflowParameters})} />
-                        )}
-                        {this.state.parameters.length > 0 && <ParametersInput parameters={this.state.parameters} onChange={parameters => this.setState({parameters})} />}
-                        {this.state.workflowParameters.length === 0 && this.state.parameters.length === 0 ? (
-                            <>
-                                <br />
-                                <label>No parameters</label>
-                            </>
-                        ) : (
-                            <></>
-                        )}
-                    </div>
-                    <div key='labels' style={{marginBottom: 25}}>
-                        <label>Labels</label>
-                        <TagsInput tags={this.state.labels} onChange={labels => this.setState({labels})} />
-                    </div>
-                    <div key='submit'>
-                        <button onClick={() => this.submit()} className='argo-button argo-button--base' disabled={this.state.isSubmitting}>
-                            <i className='fa fa-plus' /> {this.state.isSubmitting ? 'Loading...' : 'Submit'}
-                        </button>
-                    </div>
-                </div>
-            </>
-        );
-    }
+    const [entrypoint, setEntrypoint] = useState<string>(workflowEntrypoint);
+    const [parameters, setParameters] = useState<Parameter[]>([]);
+    const [workflowParameters, setWorkflowParameters] = useState<Parameter[]>(JSON.parse(JSON.stringify(props.workflowParameters)));
+    const [templates] = useState<Template[]>([defaultTemplate].concat(props.templates));
```

Comment:
Originally state is used, but should we change here without useState?

```
 const templates = [defaultTemplate].concat(props.templates);
```

Reply:
Yea this doesn't seem like it actually uses state, just derives.
If that was meant to be a performance optimization, `useMemo` should be used instead (I believe there's a few places in the code where it seemed like it might be more appropriate for `useMemo`).

Reply:
I changed `templates` from state to const and memoized the select options created based on 'templates'
Could you review if the changes are appropriate?
https://github.com/argoproj/argo-workflows/pull/11803/commits/9050132b89d453d9faaef5ce77af15590bee6a2d

Reply:
I think the memoization might not work as you're comparing `templates`, which is an array. `useMemo` [uses `Object.is`](https://react.dev/reference/react/useMemo#skipping-expensive-recalculations) to check equality of the dependencies, which uses [reference equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is#description) for objects (and arrays). So it will probably get recalculated every time as `templates` is recalculated it above it every time.

I'm fine with no `useMemo` and just having it be a normal variable, as the performance is not really significant. Otherwise can try using `props.templates` as the dependency instead (as `templates` is wholly derived from `props.templates`, they are semantically equivalent dependencies), which I _think_ would pass reference equality checks as it [does not change after the initial fetch of the `WorkflowTemplate`](https://github.com/argoproj/argo-workflows/blob/55bb51885d2a6690727f97dce25fffef1afb34f2/ui/src/app/workflow-templates/components/workflow-template-details/workflow-template-details.tsx#L133)

Reply:
thanks for elaborating on useMemo!
I now have a better understanding. (I didn't understand it before😃)
I fixed it and confirmed that variables templates and templatesOptions are not recalculated by using console.log.

Reply:
Glad to help 🙂
New version looks correct to me!